File: fxgl-core/src/main/java/com/almasb/fxgl/core/concurrent/IOTask.java
/*
 * FXGL - JavaFX Game Library. The MIT License (MIT).
 * Copyright (c) AlmasB (almaslvl@gmail.com).
 * See LICENSE for details.
 */

package com.almasb.fxgl.core.concurrent;

import com.almasb.fxgl.core.util.EmptyRunnable;
import com.almasb.fxgl.logging.Logger;
import javafx.concurrent.Task;

import java.util.concurrent.Callable;
import java.util.function.Consumer;
import java.util.function.Function;

/**
 * IO Task that wraps some IO or any other operation
 * that may potentially fail.
 *
 * @author Almas Baimagambetov (almaslvl@gmail.com)
 */
public abstract class IOTask<T> {

    private static final Logger log = Logger.get(IOTask.class);

    private static final String DEFAULT_NAME = "NoName";

    private final String name;

    private Consumer<T> successAction = (result) -> {};
    private Consumer<Throwable> failAction = (e) -> { log.warning(getName() + " failed", e); };
    private Runnable cancelAction = EmptyRunnable.INSTANCE;

    private boolean hasFailAction = false;

    private boolean isCancelled = false;

    public IOTask() {
        this(DEFAULT_NAME);
    }

    public IOTask(String name) {
        this.name = name;
    }

    public final String getName() {
        return name;
    }

    public final boolean hasFailAction() {
        return hasFailAction;
    }

    /**
     * @return whether this task was cancelled by calling "cancel()"
     */
    public boolean isCancelled() {
        return isCancelled;
    }

    /**
     * Cancel the execution of this task.
     * No-op if the task already completed.
     * It is up to each individual task to decide what to do when cancelled.
     * Tasks are free to ignore cancellation and result in onSuccess or onFailure as normal.
     * The tasks that honor cancellation must call throwCancelException(), which will be caught
     * and cancelAction invoked.
     */
    public void cancel() {
        isCancelled = true;
    }

    /**
     * Set consumer action for success scenario.
     *
     * @param successAction action to call if the task succeeds
     */
    public final IOTask<T> onSuccess(Consumer<T> successAction) {
        this.successAction = successAction;
        return this;
    }

    /**
     * Set error consumer for fail scenario.
     *
     * @param failAction exception handler to call if the task fails
     */
    public final IOTask<T> onFailure(Consumer<Throwable> failAction) {
        hasFailAction = true;
        this.failAction = failAction;
        return this;
    }

    /**
     * Set action to call on cancel.
     */
    public final IOTask<T> onCancel(Runnable cancelAction) {
        this.cancelAction = cancelAction;
        return this;
    }

    protected abstract T onExecute() throws Exception;

    protected void throwCancelException() {
        throw new IOTaskCancelledException();
    }

    /**
     * Executes this task synchronously on this thread.
     * All callbacks will be called on this thread.
     */
    public final T run() {
        try {
            if (isCancelled)
                throwCancelException();

            T value = onExecute();
            succeed(value);
            return value;
        } catch (Exception e) {
            fail(e);
            return null;
        }
    }

    /**
     * Allows chaining IO tasks to be executed sequentially.
     *
     * @param mapper function that takes result of previous task and returns new task
     * @return IO task
     */
    public final <R> IOTask<R> then(Function<T, IOTask<R>> mapper) {
        return of(name, () -> mapper.apply(onExecute()).onExecute());
    }

    public final <R> IOTask<R> thenWrap(Function<T, R> mapper) {
        return then(t -> of(() -> mapper.apply(t)));
    }

    public final Task<T> toJavaFXTask() {
        return new Task<T>() {
            @Override
            protected T call() throws Exception {
                if (isCancelled)
                    throwCancelException();

                return onExecute();
            }

            @Override
            protected void succeeded() {
                succeed(getValue());
            }

            @Override
            protected void failed() {
                fail(getException());
            }
        };
    }

    private void succeed(T result) {
        successAction.accept(result);
    }

    private void fail(Throwable error) {
        if (error instanceof IOTaskCancelledException) {
            cancelAction.run();
            return;
        }

        failAction.accept(error);
    }

    /* Static convenience methods */

    public static IOTask<Void> ofVoid(Runnable action) {
        return ofVoid(DEFAULT_NAME, action);
    }

    public static IOTask<Void> ofVoid(String name, Runnable action) {
        return of(name, () -> {
            action.run();
            return null;
        });
    }

    public static <R> IOTask<R> of(Callable<R> action) {
        return of(DEFAULT_NAME, action);
    }

    public static <R> IOTask<R> of(String name, Callable<R> action) {
        return new IOTask<R>(name) {
            @Override
            protected R onExecute() throws Exception {
                return action.call();
            }
        };
    }

    private static class IOTaskCancelledException extends RuntimeException {

        IOTaskCancelledException() {
            super("IOTask was cancelled with cancel()");
        }
    }
}

File: fxgl/src/main/kotlin/com/almasb/fxgl/io/IOTask.kt
/*
 * FXGL - JavaFX Game Library. The MIT License (MIT).
 * Copyright (c) AlmasB (almaslvl@gmail.com).
 * See LICENSE for details.
 */

package com.almasb.fxgl.io

import com.almasb.fxgl.app.FXGL
import com.almasb.fxgl.core.logging.Logger
import javafx.concurrent.Task
import java.util.concurrent.Callable
import java.util.concurrent.Executor
import java.util.function.Consumer

/**
 * IO Task that wraps some IO or any other operation
 * that may potentially fail.
 *
 * @param T type of the result of this task
 * @author Almas Baimagambetov (almaslvl@gmail.com)
 */
abstract class IOTask<T>
@JvmOverloads constructor(val name: String = "NoName") {

    companion object {
        private val log = Logger.get(IOTask::class.java)

        /* Convenient way of creating small tasks in Java */

        @JvmStatic fun <R> of(func: Callable<R>) = of("NoName", func)

        @JvmStatic fun <R> of(taskName: String, func: Callable<R>) = object : IOTask<R>(taskName) {
            override fun onExecute(): R {
                return func.call()
            }
        }

        @JvmStatic fun ofVoid(func: Runnable) = ofVoid("NoName", func)

        @JvmStatic fun ofVoid(taskName: String, func: Runnable) = object : IOTask<Void?>(taskName) {
            override fun onExecute(): Void? {
                func.run()
                return null
            }
        }
    }

    private var onSuccess: ((T) -> Unit)? = null
    private var onFailure: ((Throwable) -> Unit)? = null

    @Throws(Exception::class)
    protected abstract fun onExecute(): T

    /**
     * Set consumer action for success scenario.
     * Note: the consumer will be invoked on the same thread
     * that invoked execute(), so be careful if your code
     * needs to be run on a particular thread.
     * See [executeAsyncWithDialogFX] for JavaFX.
     *
     * @param action action to call if the task succeeds
     */
    fun onSuccess(action: Consumer<T>): IOTask<T> {
        onSuccess = { action.accept(it) }
        return this
    }

    fun onSuccessKt(action: (T) -> Unit): IOTask<T> {
        onSuccess = action
        return this
    }

    /**
     * Set error consumer for fail scenario.
     * Note: the consumer will be invoked on the same thread
     * that invoked execute(), so be careful if your code
     * needs to be run on a particular thread.
     * See [executeAsyncWithDialogFX] for JavaFX.
     *
     * @param handler exception handler to call if the task fails
     */
    fun onFailure(handler: Consumer<Throwable>): IOTask<T> {
        onFailure = { handler.accept(it) }
        return this
    }

    fun onFailureKt(handler: (Throwable) -> Unit): IOTask<T> {
        onFailure = handler
        return this
    }

    open protected fun succeed(value: T) {
        log.debug("Task succeeded: $name")
        onSuccess?.invoke(value)
    }

    open protected fun fail(error: Throwable) {
        log.warning("Task failed: $name Error: $error")
        if (onFailure == null) {
            // https://github.com/AlmasB/FXGL/issues/480
            FXGL.getExceptionHandler().handle(error)
        } else {
            onFailure!!.invoke(error)
        }
    }

    /**
     * Allows chaining IO tasks to be executed sequentially.
     *
     * @param mapper function that takes result of previous task and returns new task
     * @return IO task
     */
    fun <R> then(mapper: (T) -> IOTask<R>) = taskOf(name, {

        log.debug("Executing task: $name")

        val result = this@IOTask.onExecute()

        val otherTask = mapper.invoke(result)

        log.debug("Executing task: ${otherTask.name}")

        return@taskOf otherTask.onExecute()
    })

    /**
     * Executes this task synchronously.
     * Note: "Void" tasks will also return null.
     *
     * @return value IO task resulted in, or null if failed
     */
    fun execute(): T? {
        log.debug("Executing task: $name")

        try {
            val value = onExecute()
            succeed(value)
            return value
        } catch (e: Exception) {
            fail(e)
            return null
        }
    }

    /**
     * Executes this task asynchronously with given executor.
     * Note: it is up to the caller to ensure that executor is actually async.
     *
     * @param executor executor to use for async
     */
    @JvmOverloads fun executeAsync(executor: Executor = FXGL.getExecutor()) {
        executor.execute({ execute() })
    }

    /**
     * Executes this task asynchronously with given executor and shows dialog.
     * The dialog will be dismissed after task is completed, whether succeeded or failed.
     * Note: it is up to the caller to ensure that executor is actually async.
     *
     * @param executor executor to use for async
     * @param dialog dialog to use while task is being executed
     */
    @JvmOverloads fun executeAsyncWithDialogFX(dialog: UIDialogHandler = NONE, executor: Executor = FXGL.getExecutor()) {
        log.debug("Showing dialog")
        dialog.show()

        val task = object : Task<T>() {
            override fun call(): T {
                log.debug("Executing task: $name")
                return onExecute()
            }

            override fun succeeded() {
                log.debug("succeed(): Dismissing dialog")
                dialog.dismiss()
                succeed(value)
            }

            override fun failed() {
                log.debug("fail(): Dismissing dialog")
                dialog.dismiss()
                fail(exception)
            }
        }

        executor.execute(task)
    }
}

fun <R> taskOf(func: () -> R) = taskOf("NoName", func)

fun <R> taskOf(taskName: String, func: () -> R): IOTask<R> = object : IOTask<R>(taskName) {
    override fun onExecute(): R {
        return func()
    }
}

fun <R> voidTaskOf(func: () -> R) = voidTaskOf("NoName", func)

fun <R> voidTaskOf(taskName: String, func: () -> R) = object : IOTask<Void?>(taskName) {
    override fun onExecute(): Void? {
        func()
        return null
    }
}